### PR TITLE
WASM: Fast array implementation

### LINKED ISSUE #1888: WASM - fast array out of bounds traps
Use the fast array implementation to detect out of bounds access and trap.

### PR DESCRIPTION
**Pull request by Cellule**
Closes #1888
Related #2100

This PR implements the fast array (aka 8GB array) for WebAssembly on Windows x64.
I have put the feature under a flag (turned on by default) to help with testing and possibly debugging future issues.
This creates a new type of ArrayBuffer, `WebAssemblyArrayBuffer ` which, since the creation is controlled within WebAssembly, always allocate the buffer in virtual memory regardless of size or asm.js constraint.

Performance data from the console version of the unity benchmark
```
unity                  Left score        Right score       ∆ Score   ∆ Score %  Comment
---------------------  ----------------  ----------------  --------  ---------  --------
Overall                 63168.42 ±0.12%   73064.71 ±0.12%   9896.30     15.67%  Improved
Mandelbrot Script       68003.40 ±0.14%   68933.23 ±0.17%    929.83      1.37%  Improved
Instantiate & Destroy   20400.44 ±0.11%   23666.88 ±0.07%   3266.43     16.01%  Improved
CryptoHash Script      130857.13 ±0.18%  160959.62 ±0.07%  30102.49     23.00%  Improved
Animation & Skinning      394.56 ±0.20%     454.41 ±0.33%     59.86     15.17%  Improved
Asteroid Field          16249.17 ±0.20%   19167.92 ±0.18%   2918.75     17.96%  Improved
Physics Spheres          1221.60 ±0.09%    1373.78 ±0.06%    152.18     12.46%  Improved
Physics Spheres          1221.60 ±0.09%    1373.78 ±0.06%    152.18     12.46%  Improved
2D Physics Boxes         1540.40 ±0.21%    1700.38 ±0.10%    159.98     10.39%  Improved
---------------------  ----------------  ----------------  --------  ---------  --------
Total                    8207.50 ±0.15%    9331.11 ±0.13%   1123.61     13.69%  Improved
```

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/2423)

### CODE REVIEW

**Comment by Cellule** [comment]
@Krovatkin @commonlisp @arunetm You guys might be interested in this

File: lib/Runtime/Library/ArrayBuffer.cpp [hunk 1]
```diff
@@ -834,6 +823,91 @@ namespace Js
...
+#else
+        return false;
+#endif
+    }
+
+    ArrayBuffer * WebAssemblyArrayBuffer::TransferInternal(uint32 newBufferLength)
+    {
+        if (newBufferLength == this->bufferLength)
+        {
+            return this;
```
Comment (Cellule) [inline lib/Runtime/Library/ArrayBuffer.cpp]: @MikeHolman I don't fully understand the spec for TransferInternal.
Must I always return a **new** ArrayBuffer or returning the current one is valid if the buffer doesn't change ?

**Comment by Krovatkin** [comment]
@Cellule we are very much so, indeed! Thank you very much for a heads-up. We have high hopes that this change will improve some benchmark numbers 🎆

**Comment by Cellule** [comment]
ping

File: lib/Runtime/Library/ArrayBuffer.cpp (hunk 1 above)
Comment (MikeHolman) [inline lib/Runtime/Library/ArrayBuffer.cpp]: Yes, it should always return a new ArrayBuffer

File: lib/Runtime/Library/JavascriptFunction.cpp [hunk 2]
```diff
@@ -2282,9 +2296,16 @@ void __cdecl _alloca_probe_16()
         }
 
         // If asm.js, make sure the base address is that of the heap buffer
-        if (isAsmJs && (instrData.bufferValue != (uint64)buffer))
+        if (instrData.bufferValue != (uint64)buffer)
         {
-            return false;
+            if (isAsmJs)
```
Comment (MikeHolman) [inline lib/Runtime/Library/JavascriptFunction.cpp]: i think we should make it more clear that this also will include wasm, e.g. by renaming this var

File: lib/Runtime/Library/JavascriptFunction.cpp [hunk 3]
```diff
@@ -2282,9 +2296,16 @@ void __cdecl _alloca_probe_16()
...
+            {
+                return false;
+            }
+        }
+        else if (isWasm)
+        {
+            JavascriptError::ThrowWebAssemblyRuntimeError(func->GetScriptContext(), JSERR_InvalidTypedArrayIndex);
```
Comment (MikeHolman) [inline lib/Runtime/Library/JavascriptFunction.cpp]: gotta say, i'm kind of amazed that this "just works"

**Comment by MikeHolman** [comment]
:shipit:

**Pull request merged**